src/Development/Shake/Internal/Args.hs

```diff
@@ -169,6 +169,15 @@ shakeArgsOptionsWith baseOpts userOptions rules = do
                             xs <- getTargets shakeOpts rs
                             evaluate $ force ["  - " ++ a ++ maybe "" (" - " ++) b | (a,b) <- xs]
                         _ -> return []
+            helpSuffix <- if not long then return [] else
+                handleSynchronous (\e -> do putWhenLn Normal $ "Failure to get help suffix: " ++ show e; return []) $ do
+                    -- run the rules as simply as we can
+                    rs <- rules shakeOpts [] []
+                    case rs of
+                        Just (_, rs) -> do
+                            helpSuffix <- getHelpSuffix shakeOpts rs
```

Comment:
We only want to run the code at most once, so `getHelpSuffix` should be just after `getTargets` above. The code should look alone the lines of `(targets, helpSuffix) <- if not long then return ([], []) else ...`